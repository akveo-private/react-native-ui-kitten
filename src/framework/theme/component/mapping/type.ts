// TODO(mapping/type): declare Config/Token type

export type ThemeMappingConfigType = any;

export interface ThemeMappingType {
  parameters: string[];
  variants: any;
}

export interface VariantType {
  state: any;
}

export type TokenType = any;
