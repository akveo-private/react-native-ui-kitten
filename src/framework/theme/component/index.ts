export { ThemeProvider } from './themeProvider.component';
export { withTheme } from './themeConsumer.component';
export { withThemedStyles } from './themeConsumerStyled.component';
export { ThemeType } from './type';
