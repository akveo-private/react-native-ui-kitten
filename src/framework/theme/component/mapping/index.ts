export * from './type';
export {
  MappingProvider as ThemeMappingProvider,
  Props as ThemeMappingProviderProps,
} from './mappingProvider.component';
