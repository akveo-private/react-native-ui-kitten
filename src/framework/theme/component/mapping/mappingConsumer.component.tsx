import React, { ComponentType } from 'react';
import MappingContext from './mappingContext';
import { ThemeMappingType } from './type';
import {
  getComponentThemeMapping,
  forwardProps,
} from '../../service';

export interface Props<P> extends React.ClassAttributes<P> {
  mapping: ThemeMappingType;
}

export function withMapping<P extends Props<P>>(Component: ComponentType<P>) {
  type TExcept = Exclude<keyof P, keyof Props<P>>;
  type ForwardedProps = Pick<P, TExcept>;

  class Wrapper extends React.Component<ForwardedProps> {
    wrappedComponentRef = undefined;
    getWrappedInstance = undefined;

    setWrappedComponentRef = (ref) => {
      this.wrappedComponentRef = ref;
    };

    getComponentName = () => Component.displayName || Component.name;

    renderWrappedComponent = (mapping: ThemeMappingType[]) => (
      <Component
        ref={this.setWrappedComponentRef}
        mapping={getComponentThemeMapping(this.getComponentName(), mapping)}
        {...this.props}
      />
    );

    render() {
      return (
        <MappingContext.Consumer>
          {this.renderWrappedComponent}
        </MappingContext.Consumer>
      );
    }
  }

  const Result = Wrapper;
  Result.prototype.getWrappedInstance = function getWrappedInstance() {
    const hasWrappedInstance = this.wrappedComponentRef && this.wrappedComponentRef.getWrappedInstance;
    return hasWrappedInstance ? this.wrappedComponentRef.getWrappedInstance() : this.wrappedComponentRef;
  };

  return forwardProps(Component, Result);
}
