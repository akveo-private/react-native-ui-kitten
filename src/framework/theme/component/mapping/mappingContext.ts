import React from 'react';
import { ThemeMappingType } from './type';

const defaultValue: ThemeMappingType = {};
const MappingContext = React.createContext(defaultValue);

export default MappingContext;
