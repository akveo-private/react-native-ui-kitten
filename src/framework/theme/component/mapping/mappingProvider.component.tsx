import React from 'react';
import MappingContext from './mappingContext';
import { ThemeMappingType } from './type';

export interface Props {
  mapping: ThemeMappingType;
  children: JSX.Element | React.ReactNode;
}

export class MappingProvider extends React.PureComponent<Props> {

  render() {
    return (
      <MappingContext.Provider
        value={this.props.mapping}>
        {this.props.children}
      </MappingContext.Provider>
    );
  }
}
