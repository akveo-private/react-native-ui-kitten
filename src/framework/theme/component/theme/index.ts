export * from './type';
export {
  ThemeProvider,
  Props as ThemeProviderProps,
} from './themeProvider.component';
export {
  withStyles,
  Props as ThemedComponentProps,
} from './themeConsumer.component';
