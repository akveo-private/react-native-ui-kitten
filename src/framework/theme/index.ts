export {
  StyleProvider,
  styled,
  ThemeProvider,
  withStyles,

  StyleProviderProps,
  StyledComponentProps,
  ThemeProviderProps,
  ThemedComponentProps,

  ThemeType,
  ThemeMappingType,
  ComponentMappingType,
  TokenType,
  ThemedStyleType,
  StyleSheetType,
  StyleType,
} from './component';

export {
  APPEARANCE_DEFAULT,
} from './service';

