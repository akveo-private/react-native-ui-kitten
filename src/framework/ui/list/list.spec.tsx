import React from 'react';
import {
  GestureResponderEvent,
  Image,
  ImageProps,
  ImageSourcePropType,
  TouchableOpacity,
} from 'react-native';
import {
  render,
  fireEvent,
  shallow,
  RenderAPI,
} from 'react-native-testing-library';
import { ReactTestInstance } from 'react-test-renderer';
import {
  styled,
  ApplicationProvider,
  ApplicationProviderProps,
  StyleType,
} from '@kitten/theme';
import {
  List as ListComponent,
  Props as ListProps,
} from './list.component';
import {
  ListItem as ListItemComponent,
  Props as ListItemProps,
} from './listItem.component';
import { default as mapping } from '../common/mapping.json';
import { default as theme } from '../common/theme.json';

const List = styled<ListProps>(ListComponent);
const ListItem = styled<ListItemProps>(ListItemComponent);

const data: any[] = Array(8);

const Mock = (props?: ListProps): React.ReactElement<ApplicationProviderProps> => {
  return (
    <ApplicationProvider
      mapping={mapping}
      theme={theme}>
      <List {...props} />
    </ApplicationProvider>
  );
};

const ItemMock = (props?: ListItemProps): React.ReactElement<ListItemProps> => {
  return (
    <ListItem {...props} />
  );
};

describe('@list: component checks', () => {

  it('* renders proper amount of data', () => {
    const item = () => {
      return (
        <ItemMock title='Title'/>
      );
    };

    const component: RenderAPI = render(
      <Mock
        data={data}
        renderItem={item}
      />,
    );

    const items: ReactTestInstance[] = component.getAllByType(ListItemComponent);

    expect(items.length).toEqual(8);
  });

});

describe('@list-item: template matches snapshot', () => {

  const iconSource: ImageSourcePropType = { uri: 'https://akveo.github.io/eva-icons/fill/png/128/star.png' };

  it('* title', () => {
    const item = () => {
      return (
        <ItemMock title='Title'/>
      );
    };

    const component: RenderAPI = render(
      <Mock
        data={data}
        renderItem={item}
      />,
    );

    const items: ReactTestInstance[] = component.getAllByType(ListItemComponent);
    const { output } = shallow(items[0]);

    expect(output).toMatchSnapshot();
  });

  it('* description', () => {
    const item = () => {
      return (
        <ItemMock description='Description'/>
      );
    };

    const component: RenderAPI = render(
      <Mock
        data={data}
        renderItem={item}
      />,
    );

    const items: ReactTestInstance[] = component.getAllByType(ListItemComponent);
    const { output } = shallow(items[0]);

    expect(output).toMatchSnapshot();
  });

  it('* text styles', () => {
    const item = () => {
      return (
        <ItemMock
          title='Title'
          titleStyle={{ fontSize: 22, lineHeight: 24 }}
          description='Description'
          descriptionStyle={{ color: 'blue', letterSpacing: 6 }}
        />
      );
    };

    const component: RenderAPI = render(
      <Mock
        data={data}
        renderItem={item}
      />,
    );

    const items: ReactTestInstance[] = component.getAllByType(ListItemComponent);
    const { output } = shallow(items[0]);

    expect(output).toMatchSnapshot();
  });

  it('* with icon', () => {
    const item = () => {
      return (
        <ItemMock
          title='Title'
          description='Description'
          icon={icon}
        />
      );
    };

    const icon = (style: StyleType): React.ReactElement<ImageProps> => {
      return (
        <Image
          style={style}
          source={iconSource}
        />
      );
    };

    const component: RenderAPI = render(
      <Mock
        data={data}
        renderItem={item}
      />,
    );

    const items: ReactTestInstance[] = component.getAllByType(ListItemComponent);
    const { output } = shallow(items[0]);

    expect(output).toMatchSnapshot();
  });

  it('* with accessory', () => {
    const item = () => {
      return (
        <ItemMock
          title='Title'
          description='Description'
          accessory={accessory}
        />
      );
    };

    const accessory = (style: StyleType): React.ReactElement<ImageProps> => {
      return (
        <Image
          style={style}
          source={iconSource}
        />
      );
    };

    const component: RenderAPI = render(
      <Mock
        data={data}
        renderItem={item}
      />,
    );

    const items: ReactTestInstance[] = component.getAllByType(ListItemComponent);
    const { output } = shallow(items[0]);

    expect(output).toMatchSnapshot();
  });

});

describe('@list-item: component checks', () => {

  it('* emits onPress with correct args', async () => {
    const pressIndex: number = 0;

    const onPress = jest.fn((index: number) => {
      expect(index).toEqual(pressIndex);
    });

    const item = () => {
      return (
        <ItemMock
          title='Title'
          onPress={onPress}
        />
      );
    };

    const component: RenderAPI = render(
      <Mock
        data={data}
        renderItem={item}
      />,
    );

    const items: ReactTestInstance[] = component.getAllByType(ListItemComponent);
    const touchable: ReactTestInstance = items[pressIndex].findByType(TouchableOpacity);

    fireEvent.press(touchable);
  });

});
