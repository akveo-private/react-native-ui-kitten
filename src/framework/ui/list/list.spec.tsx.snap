// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@list-item: template matches snapshot * description 1`] = `
<TouchableOpacity
  activeOpacity={1}
  appearance="default"
  dispatch={[Function]}
  index={0}
  onLongPress={[Function]}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    Array [
      Object {
        "paddingHorizontal": 6,
        "paddingVertical": 12,
      },
      Array [
        undefined,
        undefined,
        Object {},
      ],
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
  title="Title"
>
  <View
    style={
      Array [
        Object {
          "accessory": Object {
            "height": 48,
            "marginHorizontal": 6,
            "width": 48,
          },
          "description": Object {
            "color": "#a6aebd",
            "fontSize": 14,
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          "icon": Object {
            "height": 48,
            "marginHorizontal": 6,
            "tintColor": "#A6AEBD",
            "width": 48,
          },
          "title": Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <ForwardRef(WrappingElement)
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          Object {},
        ]
      }
    >
      Title
    </ForwardRef(WrappingElement)>
  </View>
</TouchableOpacity>
`;

exports[`@list-item: template matches snapshot * title 1`] = `
<TouchableOpacity
  activeOpacity={1}
  appearance="default"
  dispatch={[Function]}
  index={0}
  onLongPress={[Function]}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    Array [
      Object {
        "paddingHorizontal": 6,
        "paddingVertical": 12,
      },
      Array [
        undefined,
        undefined,
        Object {},
      ],
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
  title="Title"
>
  <View
    style={
      Array [
        Object {
          "accessory": Object {
            "height": 48,
            "marginHorizontal": 6,
            "width": 48,
          },
          "description": Object {
            "color": "#a6aebd",
            "fontSize": 14,
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          "icon": Object {
            "height": 48,
            "marginHorizontal": 6,
            "tintColor": "#A6AEBD",
            "width": 48,
          },
          "title": Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <ForwardRef(WrappingElement)
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          Object {},
        ]
      }
    >
      Title
    </ForwardRef(WrappingElement)>
  </View>
</TouchableOpacity>
`;

exports[`@list-item: template matches snapshot * with accessory 1`] = `
<TouchableOpacity
  accessory={[Function]}
  activeOpacity={1}
  appearance="default"
  description="Description"
  dispatch={[Function]}
  index={0}
  onLongPress={[Function]}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    Array [
      Object {
        "paddingHorizontal": 6,
        "paddingVertical": 12,
      },
      Array [
        undefined,
        undefined,
        Object {},
      ],
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
  title="Title"
>
  <View
    style={
      Array [
        Object {
          "accessory": Object {
            "height": 48,
            "marginHorizontal": 6,
            "width": 48,
          },
          "description": Object {
            "color": "#a6aebd",
            "fontSize": 14,
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          "icon": Object {
            "height": 48,
            "marginHorizontal": 6,
            "tintColor": "#A6AEBD",
            "width": 48,
          },
          "title": Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <ForwardRef(WrappingElement)
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          Object {},
        ]
      }
    >
      Title
    </ForwardRef(WrappingElement)>
    <ForwardRef(WrappingElement)
      style={
        Array [
          Object {
            "color": "#a6aebd",
            "fontSize": 14,
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          Object {},
        ]
      }
    >
      Description
    </ForwardRef(WrappingElement)>
  </View>
  <Component
    source={
      Object {
        "uri": "https://akveo.github.io/eva-icons/fill/png/128/star.png",
      }
    }
    style={
      Array [
        0,
        Object {
          "height": 48,
          "marginHorizontal": 6,
          "width": 48,
        },
        Object {},
      ]
    }
  />
</TouchableOpacity>
`;

exports[`@list-item: template matches snapshot * with icon 1`] = `
<TouchableOpacity
  activeOpacity={1}
  appearance="default"
  description="Description"
  dispatch={[Function]}
  icon={[Function]}
  index={0}
  onLongPress={[Function]}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    Array [
      Object {
        "paddingHorizontal": 6,
        "paddingVertical": 12,
      },
      Array [
        undefined,
        undefined,
        Object {},
      ],
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
  title="Title"
>
  <Component
    source={
      Object {
        "uri": "https://akveo.github.io/eva-icons/fill/png/128/star.png",
      }
    }
    style={
      Array [
        Object {
          "height": 48,
          "marginHorizontal": 6,
          "tintColor": "#A6AEBD",
          "width": 48,
        },
        0,
        Object {},
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "accessory": Object {
            "height": 48,
            "marginHorizontal": 6,
            "width": 48,
          },
          "description": Object {
            "color": "#a6aebd",
            "fontSize": 14,
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          "icon": Object {
            "height": 48,
            "marginHorizontal": 6,
            "tintColor": "#A6AEBD",
            "width": 48,
          },
          "title": Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <ForwardRef(WrappingElement)
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          Object {},
        ]
      }
    >
      Title
    </ForwardRef(WrappingElement)>
    <ForwardRef(WrappingElement)
      style={
        Array [
          Object {
            "color": "#a6aebd",
            "fontSize": 14,
            "marginHorizontal": 6,
            "marginVertical": 4,
          },
          Object {},
        ]
      }
    >
      Description
    </ForwardRef(WrappingElement)>
  </View>
</TouchableOpacity>
`;
